Assignment4/Rupesh_punna_X3123008/task07.py

```diff
@@ -0,0 +1,162 @@
+# -*- coding: utf-8 -*-
+"""Task07.ipynb
+
+Automatically generated by Colab.
+
+Original file is located at
+    https://colab.research.google.com/drive/1OOVwQspi3ec1dG0WGJauGbYpYCJctvqx
+
+**Task 07: Querying RDF(s)**
+"""
+
+#!pip install rdflib
+import urllib.request
+url = 'https://raw.githubusercontent.com/FacultadInformatica-LinkedData/Curso2025-2026/refs/heads/master/Assignment4/course_materials/python/validation.py'
+urllib.request.urlretrieve(url, 'validation.py')
+github_storage = "https://raw.githubusercontent.com/FacultadInformatica-LinkedData/Curso2025-2026/master/Assignment4/course_materials"
+
+from validation import Report
+
+"""First let's read the RDF file"""
+
+from rdflib import Graph, Namespace, Literal
+from rdflib.namespace import RDF, RDFS
+# Do not change the name of the variables
+g = Graph()
+g.namespace_manager.bind('ns', Namespace("http://somewhere#"), override=False)
+g.parse(github_storage+"/rdf/data06.ttl", format="TTL")
+report = Report()
+
+"""**TASK 7.1a: For all classes, list each classURI. If the class belogs to another class, then list its superclass.**
+**Do the exercise in RDFLib returning a list of Tuples: (class, superclass) called "result". If a class does not have a super class, then return None as the superclass**
+"""
+
+result = []
+for c in g.subjects(RDF.type, RDFS.Class):
+  superclass = g.value(subject=c, predicate=RDFS.subClassOf, default=None)
+  result.append((c, superclass))
+
+# Visualize the results
+for r in result:
+  print(r)
+
+## Validation: Do not remove
+report.validate_07_1a(result)
+
+"""**TASK 7.1b: Repeat the same exercise in SPARQL, returning the variables ?c (class) and ?sc (superclass)**"""
+
+query =  """
+SELECT DISTINCT ?c ?sc
+WHERE {
+  ?c a rdfs:Class .
+  OPTIONAL { ?c rdfs:subClassOf ?sc }
+}
+"""
+
+for r in g.query(query):
+  print(r.c, r.sc)
+
+## Validation: Do not remove
+report.validate_07_1b(query,g)
+
+"""**TASK 7.2a: List all individuals of "Person" with RDFLib (remember the subClasses). Return the individual URIs in a list called "individuals"**
+
+"""
+
+ns = Namespace("http://oeg.fi.upm.es/def/people#")
+individuals = []
+
+person_classes = set([ns.Person])
+for s, p, o in g.triples((None, RDFS.subClassOf, ns.Person)):
+  person_classes.add(s)
+
+person_classes.add(ns.Professor)
+person_classes.add(ns.FullProfessor)
+person_classes.add(ns.AssociateProfessor)
+person_classes.add(ns.InterimAssociateProfessor)
+
+for cls in person_classes:
+  for individual in g.subjects(RDF.type, cls):
+    individuals.append(individual)
+
+individuals = sorted(list(set(individuals)))
```

Comment:
You can't list the full class hierarchy directly. You need to use the graph in rdf only